api/ocm/compdesc/resolver.go

```diff
@@ -0,0 +1,96 @@
+package compdesc
+
+import (
+	"sync"
+
+	"github.com/mandelsoft/goutils/errors"
+
+	common "ocm.software/ocm/api/utils/misc"
+)
+
+type ComponentVersionResolver interface {
+	LookupComponentVersion(name string, version string) (*ComponentDescriptor, error)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+type ComponentVersionSet struct {
+	lock sync.RWMutex
+	cds  map[common.NameVersion]*ComponentDescriptor
+}
+
+var _ ComponentVersionResolver = (*ComponentVersionSet)(nil)
+
+func NewComponentVersionSet(cds ...*ComponentDescriptor) *ComponentVersionSet {
+	r := map[common.NameVersion]*ComponentDescriptor{}
+	for _, cd := range cds {
+		r[common.NewNameVersion(cd.Name, cd.Version)] = cd.Copy()
+	}
+	return &ComponentVersionSet{cds: r}
+}
+
+func (c *ComponentVersionSet) LookupComponentVersion(name string, version string) (*ComponentDescriptor, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+
+	nv := common.NewNameVersion(name, version)
+
+	cd := c.cds[nv]
+	if cd == nil {
+		return nil, errors.ErrNotFound(KIND_COMPONENTVERSION, nv.String())
+	}
+	return cd, nil
+}
+
+func (c *ComponentVersionSet) AddVersion(cd *ComponentDescriptor) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	c.cds[common.NewNameVersion(cd.Name, cd.Version)] = cd.Copy()
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+type CompoundResolver struct {
+	lock      sync.RWMutex
+	resolvers []ComponentVersionResolver
+}
+
+var _ ComponentVersionResolver = (*CompoundResolver)(nil)
+
+func NewCompoundResolver(res ...ComponentVersionResolver) ComponentVersionResolver {
+	for i := 0; i < len(res); i++ {
+		if res[i] == nil {
+			res = append(res[:i], res[i+1:]...)
+			i--
+		}
+	}
+	if len(res) == 1 {
```

Comment:
Are you anticipating that someone might initialize this resolver with `(res, res, nil, nil, res, res)`? How would that be possible, or why would that be possible?

Reply:
the resolver list is typically variadic. On other places you may pass a resolver, which could be nil.
Therefore it seems to be useful to filter the nil at the final place , nit at all places where it is called to avoid always code
like 
```
   if resolver == nil {
       call(...)
   }  else {
       call(..., resolver)
   }
```

And generally this is some kind of defensive programming.

Reply:
But if you don't pass in anything the slice is just plain empty so nothing would happen. You would have to explicitly pass in a `nil` for that to be there something or something kind of list where there are nils in-between because the resolver was somehow not initialized.

And yes, I agree that you need to do the defence, but you are actually already doing that further down when you are looping through these resolvers here:

```go
	for _, r := range c.resolvers {
		if r == nil {
			continue
		}
```

So I think this looping here isn't really necessary.

Reply:
yes, sure, but when creating a compound resolver with one effective sub resolver the useless compound feature can be omitted, instead, the sole resolver is returned.
This requires to eliminate the nil resolvers first. What could be done is to omit the `continue` case in the Lookup method, if really all constructors assure that there is no nil entry. So far, this is the case, but possible constructors can be added later without knowing this special constraint, when just keep this check in the Lookup implementation.

Reply:
By the way, there is already such a case in the actual coding.
The AddResolver does not check for nil.
This should be fixed.

Reply:
I've added the check

Reply:
Nice.